src/main/java/com/powsybl/openloadflow/network/LfZeroImpedanceNetwork.java

```diff
@@ -0,0 +1,212 @@
+            if (voltageControls.size() > 1) {
+                // we take the highest target voltage (why not...) and in case of equality the voltage control
+                // with the first controlled bus ID by alpha sort
+                voltageControls.sort(Comparator.<VoltageControl<?>>comparingDouble(VoltageControl::getTargetValue)
+                        .reversed()
+                        .thenComparing(o -> o.getControlledBus().getId()));
+                VoltageControl<?> mainVc = voltageControls.get(0);
+                // first one is main, the other have are dependents
+                for (int i = 1; i < voltageControls.size(); i++) {
+                    VoltageControl<?> vc = voltageControls.get(i);
+                    vc.mergeStatus = VoltageControl.MergeStatus.DEPENDENT;
+                    linkVoltageControls(mainVc, vc);
+                }
+            }
+        }
+    }
+
+    public void removeBranchAndTryToSplit(LfBranch disabledBranch) {
+        graph.removeEdge(disabledBranch);
+
+        List<Set<LfBus>> connectedSets = new ConnectivityInspector<>(graph).connectedSets();
+        if (connectedSets.size() > 1) { // real split
+            disabledBranch.setSpanningTreeEdge(dc, false);
+
+            Set<LfZeroImpedanceNetwork> zeroImpedanceNetworks = network.getZeroImpedanceNetworks(dc);
+            zeroImpedanceNetworks.remove(this);
+            List<LfZeroImpedanceNetwork> splitZns = new ArrayList<>(2);
+            for (Set<LfBus> connectedSet : connectedSets) {
+                var subGraph = new AsSubgraph<>(graph, connectedSet);
+                splitZns.add(new LfZeroImpedanceNetwork(network, dc, subGraph));
+            }
+            zeroImpedanceNetworks.addAll(splitZns);
+
+            // update voltage control merge status
+            if (!dc) {
+                for (LfZeroImpedanceNetwork splitZn : splitZns) {
+                    splitZn.updateVoltageControlMergeStatus();
+                }
+            }
+
+            for (LfNetworkListener listener : network.getListeners()) {
+                listener.onZeroImpedanceNetworkSplit(this, splitZns, dc);
+            }
+        } else {
+            if (disabledBranch.isSpanningTreeEdge(dc)) {
+                disabledBranch.setSpanningTreeEdge(dc, false);
+
+                // just update the spanning
+                updateSpanningTree();
+            }
+        }
+    }
+
+    public static void addBranchAndMerge(LfZeroImpedanceNetwork zn1, LfZeroImpedanceNetwork zn2, LfBranch enabledBranch) {
+        Objects.requireNonNull(zn1);
+        Objects.requireNonNull(zn2);
+        Objects.requireNonNull(enabledBranch);
+        LfNetwork network = zn1.getNetwork();
+        boolean dc = zn1.isDc();
+        Set<LfZeroImpedanceNetwork> zeroImpedanceNetworks = network.getZeroImpedanceNetworks(dc);
+        Graph<LfBus, LfBranch> mergedGraph = new Pseudograph<>(LfBranch.class);
```

Comment:
This is a new `Graph` and not a `AsSubgraph` anymore, so we're mixing subgraphs of `createZeroImpedanceSubGraph` and graphs. But I don't see any problems there as we don't play with the parent graph, we don't even retain it explicitly.

Reply:
I replaced by a subgraph + copy